File: CSMApps/Vertebrae/v1.0.0/src/main/java/com/carrpod/vertebrae/service/SessionManagerService.java
package com.carrpod.vertebrae.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.carrpod.vertebrae.R;
import com.carrpod.vertebrae.model.KilosSession;
import com.carrpod.vertebrae.network.WebSocketManager;
import com.carrpod.vertebrae.storage.SessionStorageManager;

public class SessionManagerService extends Service {

    private static final String TAG = "SessionManagerService";

    private SessionStorageManager storage;
    private WebSocketManager wsManager;
    private Handler handler = new Handler(Looper.getMainLooper());
    private Runnable reconnectRunnable;

    @Override
    public void onCreate() {
        super.onCreate();
        storage = SessionStorageManager.getInstance();
        wsManager = WebSocketManager.getInstance();

        createNotificationChannel();
        startForegroundService();
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    "vertebrae_sessions",
                    getString(R.string.notification_channel_name),
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription(getString(R.string.notification_channel_desc));
            channel.setShowBadge(false);
            getSystemService(NotificationManager.class).createNotificationChannel(channel);
        }
    }

    private void startForegroundService() {
        Notification notification = new NotificationCompat.Builder(this, "vertebrae_sessions")
                .setContentTitle(getString(R.string.app_name))
                .setContentText(getString(R.string.service_running))
                .setSmallIcon(R.drawable.ic_vertebrae_foreground)
                .setOngoing(true)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .setCategory(NotificationCompat.CATEGORY_SERVICE)
                .build();

        startForeground(1001, notification);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        restoreSessions();
        return START_STICKY;
    }

    private void restoreSessions() {
        for (KilosSession session : storage.loadAllSessions()) {
            if (session.isAutoReconnect() && session.getStatus() != KilosSession.SessionStatus.DISCONNECTED) {
                reconnectSession(session);
            }
        }

        // Periodic check
        reconnectRunnable = new Runnable() {
            @Override
            public void run() {
                checkSessions();
                handler.postDelayed(this, 30000); // 30 seconds
            }
        };
        handler.post(reconnectRunnable);
    }

    private void checkSessions() {
        for (KilosSession session : storage.loadAllSessions()) {
            switch (session.getStatus()) {
                case CONNECTED:
                case HEARTBEAT_ACTIVE:
                    // Check if heartbeat is stale
                    if (System.currentTimeMillis() - session.getLastHeartbeatAt() > session.getHeartbeatIntervalSeconds() * 2000L) {
                        wsManager.disconnectSession(session.getId());
                        reconnectSession(session);
                    }
                    break;
                case HEARTBEAT_FAILED:
                case ERROR:
                    if (session.isAutoReconnect()) {
                        reconnectSession(session);
                    }
                    break;
            }
        }
    }

    private void reconnectSession(KilosSession session) {
        KilosSession updated = session.copyWith(
                KilosSession.SessionStatus.RECONNECTING,
                null, null, null, null, null, null, null, null, null
        );
        storage.saveSession(updated);
        wsManager.connectSession(updated);
    }

    public void connectSession(KilosSession session) {
        wsManager.connectSession(session);
    }

    public void disconnectSession(String sessionId) {
        wsManager.disconnectSession(sessionId);
    }

    public void updateSession(KilosSession session) {
        storage.saveSession(session);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (reconnectRunnable != null) {
            handler.removeCallbacks(reconnectRunnable);
        }
        wsManager.disconnectAll();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: CSMApps/Vertebrae/v1.0.0/src/main/kotlin_backup/SessionManagerService.kt
package com.carrpod.vertebrae.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import androidx.core.app.NotificationCompat
import com.carrpod.vertebrae.R
import com.carrpod.vertebrae.comm.SessionCommunicator
import com.carrpod.vertebrae.model.KilosSession
import com.carrpod.vertebrae.network.WebSocketManager
import com.carrpod.vertebrae.storage.SessionStorageManager

class SessionManagerService : Service() {

    private lateinit var storage: SessionStorageManager
    private lateinit var wsManager: WebSocketManager
    private lateinit var comm: SessionCommunicator
    private val handler = Handler(Looper.getMainLooper())
    private var reconnectRunnable: Runnable? = null

    override fun onCreate() {
        super.onCreate()
        storage = SessionStorageManager(this)
        wsManager = WebSocketManager.getInstance()
        comm = SessionCommunicator.getInstance()

        createNotificationChannel()
        startForegroundService()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                "vertebrae_sessions",
                getString(R.string.notification_channel_name),
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = getString(R.string.notification_channel_desc)
                setShowBadge(false)
            }
            getSystemService(NotificationManager::class.java).createNotificationChannel(channel)
        }
    }

    private fun startForegroundService() {
        val notification = NotificationCompat.Builder(this, "vertebrae_sessions")
            .setContentTitle(getString(R.string.app_name))
            .setContentText(getString(R.string.service_running))
            .setSmallIcon(R.drawable.ic_vertebrae_foreground)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setCategory(NotificationCompat.CATEGORY_SERVICE)
            .build()

        startForeground(1001, notification)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        restoreSessions()
        return START_STICKY
    }

    private fun restoreSessions() {
        val sessions = storage.loadAllSessions()
        val settings = storage.loadSettings()

        sessions.forEach { session ->
            if (session.isAutoReconnect && session.status != KilosSession.SessionStatus.DISCONNECTED) {
                reconnectSession(session)
            }
        }

        // Start periodic check
        reconnectRunnable = object : Runnable {
            override fun run() {
                checkSessions()
                handler.postDelayed(this, 30000) // 30 seconds
            }
        }
        handler.post(reconnectRunnable!!)
    }

    private fun checkSessions() {
        val sessions = storage.loadAllSessions()
        sessions.forEach { session ->
            when (session.status) {
                KilosSession.SessionStatus.CONNECTED,
                KilosSession.SessionStatus.HEARTBEAT_ACTIVE -> {
                    // Check if heartbeat is stale
                    if (System.currentTimeMillis() - session.lastHeartbeatAt > session.heartbeatIntervalSeconds * 2000L) {
                        wsManager.reconnectSession(session.id)
                    }
                }
                KilosSession.SessionStatus.HEARTBEAT_FAILED,
                KilosSession.SessionStatus.ERROR -> {
                    if (session.isAutoReconnect) {
                        reconnectSession(session)
                    }
                }
            }
        }
    }

    private fun reconnectSession(session: KilosSession) {
        val updated = session.copyWith(status = KilosSession.SessionStatus.RECONNECTING)
        storage.saveSession(updated)
        wsManager.connectSession(updated)
    }

    fun connectSession(session: KilosSession) {
        wsManager.connectSession(session)
    }

    fun disconnectSession(sessionId: String) {
        wsManager.disconnectSession(sessionId)
    }

    fun updateSession(session: KilosSession) {
        storage.saveSession(session)
    }

    override fun onDestroy() {
        super.onDestroy()
        reconnectRunnable?.let { handler.removeCallbacks(it) }
        wsManager.disconnectAll()
    }

    override fun onBind(intent: Intent?): IBinder? = null
}